Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java

```diff
@@ -75,6 +77,56 @@ public ParquetFileReader(String filePath, SeekableChannelsProvider channelsProvi
         type = fromParquetSchema(fileMetaData.schema,fileMetaData.column_orders);
     }
 
+    private HashSet<String> columnsWithDictionaryUsedOnEveryDataPage = null;
+    public HashSet<String> getColumnsWithDictionaryUsedOnEveryDataPage() {
+        if (columnsWithDictionaryUsedOnEveryDataPage == null) {
+            columnsWithDictionaryUsedOnEveryDataPage = calculateColumnsWithDictionaryUsedOnEveryDataPage();
+        }
+        return columnsWithDictionaryUsedOnEveryDataPage;
+    }
+
+    private HashSet<String> calculateColumnsWithDictionaryUsedOnEveryDataPage() {
+        final HashSet<String> result = new HashSet<>(fileMetaData.getSchemaSize());
+        int rowGroupSeq = 0;
+        for (RowGroup rowGroup : fileMetaData.getRow_groups()) {
+            // On the first pass, for row group zero, we are going to add all columns to the set
+            // that satisfy the restriction.
+            // On later passes after zero, we will remove any column that does not satisfy
+            // the restriction.
+            for (ColumnChunk columnChunk : rowGroup.columns) {
+                final ColumnMetaData columnMeta = columnChunk.getMeta_data();
+                final String parquetColumnName = columnMeta.path_in_schema.get(0);
+                if (rowGroupSeq > 0 && !result.contains(parquetColumnName)) {
+                    continue;
+                }
+                if (columnMeta.encoding_stats == null) {
+                    continue;
+                }
+                boolean thisColumnInThisRowGroupDisproved = false;
+                for (PageEncodingStats encodingStat : columnMeta.encoding_stats) {
```

Comment:
So, we think this is a relatively cheap/low-cost thing to do? Do we always read the entire metadata in one go, or are we in fact causing file accesses here? I think this approach is predicated on it being "cheap enough". 

Alternatively, we could pull this idea out of column defs, move it to the column source level, and have column sources do this check when asked questions about symbol table support.

Reply:
This stuff is still in the thrift file metadata object; if we had to read it from the page preamble, that would mean seeking to the page positions, this is not doing that.